File: Data Structures & Algorithms/balanced-binary-tree/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        q = deque([root])

        def findHeight(root) -> int:
            if not root:
                return 0
            
            return 1 + max(findHeight(root.left), findHeight(root.right))

        while q:
            node = q.pop()
            if not node:
                continue

            if abs(findHeight(node.left) - findHeight(node.right)) > 1:
                return False
            
            if node.left:
                q.append(node.left)
            if node.right:
                q.append(node.right)

        return True

File: Data Structures & Algorithms/valid-binary-search-tree/submission-3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
   public:
    bool isValidBST(TreeNode* root) {
        auto valid = [&](this auto self, TreeNode* node, int l, int h) -> bool {
            if (!node) return true;
            if (node->val <= l or node->val >= h) return false;

            return self(node->left, l, node->val) and self(node->right, node->val, h);
        };

        return valid(root, -1001, 1001);
    }
};
